src/Features/CSharp/Portable/Snippets/CSharpForLoopSnippetProvider.cs

```diff
@@ -0,0 +1,179 @@
+
+namespace Microsoft.CodeAnalysis.CSharp.Snippets
+{
+    [ExportSnippetProvider(nameof(ISnippetProvider), LanguageNames.CSharp), Shared]
+    internal class CSharpForLoopSnippetProvider : AbstractForLoopSnippetProvider
+    {
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public CSharpForLoopSnippetProvider()
+        {
+        }
+
+        /// <summary>
+        /// Creates the for loop statement syntax.
+        /// Must be done in language specific file since there is no generic way to generate the syntax.
+        /// </summary>
+        protected override async Task<SyntaxNode> CreateForLoopStatementSyntaxAsync(Document document, CancellationToken cancellationToken)
+        {
+            var compilation = await document.Project.GetRequiredCompilationAsync(cancellationToken).ConfigureAwait(false);
+            var generator = SyntaxGenerator.GetGenerator(document);
+            var options = await document.GetAnalyzerConfigOptionsAsync(cancellationToken).ConfigureAwait(false);
+            var varBuiltInType = false;
+
+            if (options != null)
+            {
+                varBuiltInType = options.GetOption(CSharpCodeStyleOptions.VarForBuiltInTypes).Value;
+            }
+
+            var indexVariable = generator.Identifier("i");
+            var varIdentifier = SyntaxFactory.IdentifierName("var");
+
+            // Creating the variable declaration based on if the user has
+            // 'var for built in types' set in their editorconfig.
+            var variableDeclarationSyntax = varBuiltInType
+                ? SyntaxFactory.VariableDeclaration(varIdentifier,
+                    SyntaxFactory.SingletonSeparatedList(
+                        SyntaxFactory.VariableDeclarator(
+                            indexVariable,
+                            argumentList: null,
+                            SyntaxFactory.EqualsValueClause((ExpressionSyntax)generator.LiteralExpression(0)))))
+                : SyntaxFactory.VariableDeclaration(compilation.GetSpecialType(SpecialType.System_Int32).GenerateTypeSyntax(allowVar: false),
+                    SyntaxFactory.SingletonSeparatedList(
+                        SyntaxFactory.VariableDeclarator(
+                            indexVariable,
+                            argumentList: null,
+                            SyntaxFactory.EqualsValueClause((ExpressionSyntax)generator.LiteralExpression(0)))));
+
+            var forLoopSyntax = SyntaxFactory.ForStatement(variableDeclarationSyntax,
+                SyntaxFactory.SeparatedList<ExpressionSyntax>(),
+                (ExpressionSyntax)generator.LessThanExpression(
+                    generator.IdentifierName(indexVariable),
+                    // Using a temporary identifier name for now, could later be changed
+                    // to look for an iterable item in the scope of the insertion.
+                    generator.IdentifierName("length")),
```

Comment:
should this be some sort of LSP marker that the user can then tab through?

Reply:
what do you mean? It's a placeholder so the user can tab through and also rename?

Reply:
I guess Cyrus means we should place this into the `SnippetPlaceholder` ?